src/freenet/crypt/SHA256.java

```diff
@@ -101,9 +87,9 @@ public static MessageDigest getMessageDigest() {
 		try {
 			MessageDigest md = null;
 			synchronized(digests) {
-				int x = digests.size();
-				if(x == 0) md = null;
-				else md = digests.remove(x-1);
+				while (digests.size() > 0 && md == null) {
+					md = digests.remove(0).get();
+				}
```

Comment:
I don't think that the complexity matters here; it's meant to be small (we had 16 elements max).
We'll never get around clearing the queue of SoftReferences that are gone if I do LIFO... is it worth having a cleanup job on the ticker?
What type of non-blocking Queue do you have in mind? a LinkedList? that's a waste of memory IMHO